Add support for explicit disclosure in daml script 3

remyhaemmerle-da opened the pull request:
<!--
# Pull Request Checklist

- Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- Include appropriate tests
- Set a descriptive title and thorough description
- Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- Normal production system change, include purpose of change in description
- If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.
-->

samuel-williams-da commented on daml-script/daml3/Daml/Script/Internal.daml:
```diff
@@ -162,3 +163,7 @@ fromLedgerValue = dangerousCast
 
 toLedgerValue : a -> LedgerValue
 toLedgerValue = dangerousCast
+
+data Disclosure = Disclosure with
+    blob: Text
+  deriving (Eq, Ord)
```
Newline here, and maybe shouldn't go in `Internal`, this is just the most generised Question logic

remyhaemmerle-da replied:
Where should I put it then ?

samuel-williams-da replied:
Not sure, tempted to say Submit, or a new module

samuel-williams-da commented on daml-script/daml3/Daml/Script/Questions/Query.daml:
```diff
@@ -47,15 +47,21 @@ instance IsQuestion QueryContractId (Optional AnyTemplate) where command = "Quer
+--    convert : Optional AnyTemplate -> Optional t
+--    convert = fmap $ fromSome . fromAnyTemplate
+
+queryContractId: forall t p. (Template t, HasAgreement t, IsParties p) => HasCallStack => p -> ContractId t -> Script (Optional t)
+queryContractId p c = fmap (fmap $ fromSome . fromAnyTemplate . fst) $ queryContractId_ p c
+
+queryDisclosureContractId: forall t p. (Template t, IsParties p) => HasCallStack => p -> ContractId t -> Script (Optional Disclosure)
+queryDisclosureContractId p c = fmap (fmap  $ Disclosure . snd) $ queryContractId_ p c
```
Are are assuming someone would never want both?
I'd probably move the wrapping logic for each into a single function that gives `Optional (t, Disclosure)` then make these simple functions to peek into the data

remyhaemmerle-da replied:
Nothing against the idea. 
Just wondering if we will not better redesign all this mess before. (Eg. The names of all those fucntion are pretty messy) 
For now we have implementation you want is just not exposed (`queryContractId_`)

samuel-williams-da replied:
Okay, I'm happy if you can put up a quick ticket to redo this before daml3

samuel-williams-da commented on daml-script/daml3/Daml/Script/Questions/Submit.daml:
```diff
@@ -107,21 +113,23 @@ submitTreeMulti = submitTreeInternal
 
 -- | Alternate version of `trySubmit` that allows specifying the actAs and readAs parties.
-trySubmitMulti : HasCallStack => [Party] -> [Party] -> Commands a -> Script (Either SubmitError a)
+trySubmitMulti : HasCallStack => [Party] -> [Party] -> [Disclosure] -> Commands a -> Script (Either SubmitError a)
 trySubmitMulti = trySubmitInternal
```
This choice strays from the usual `Multi` pattern we have, I'd probably add more variants of the function for now, but we should look at improving the situation with the number of submit functions later.

remyhaemmerle-da replied:
Same as above, I am not very happy that everybody add his own variant without a clear strategy. 
This will end up as a mess that we will have to maintain.

samuel-williams-da commented on daml-script/daml3/Daml/Script.daml:
You'll need to export Disclosure here, even if without the constructors, else people can't write their own wrappers

samuel-williams-da commented on daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcLedgerClient.scala:
```diff
@@ -346,15 +355,18 @@ class GrpcLedgerClient(
       ec: ExecutionContext,
       mat: Materializer,
   ): Future[ScriptLedgerClient.TransactionTree] = {
+    assert(disclosures.isEmpty, "fix me !")
```
Why this and the noDisclosures?

remyhaemmerle-da replied:
Good catch.

samuel-williams-da commented on compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala:
```diff
@@ -356,6 +356,9 @@ final class Conversions(
         builder.setLookupError(nstBuilder.build)
+
+      case Error.DisclosureDecoding(message) =>
+        builder.setCrash("disclosure decode failed: " + message)
```
Maybe a TODO to send this over formally?

samuel-williams-da reviewed (approved):
How does this play with `trySubmit` errors? Is it possible for trySubmit to catch this, should it, and if so, could you implement that? (maybe future PR)

remyhaemmerle-da commented:
@samuel-williams-da I rebase my PR because of some conflict. 
However you can see ec144242e10f769c529fbf3dc101acadc6b3edcd where I addressed your comments.
